### ISSUE TITLE
Implement common Moodle activity events 

### ISSUE THREAD

**Issue by nadavkav** [h5p/h5p-moodle-plugin#69]
- Activity viewed
- Activity index viewed (when implemented)
- For Interactive Video: Video player events (Play, stop, seek, pause, interaction?, finished)
- Grade updated
- xAPI statement fired (maybe? instead or in parallel of sending it to an external LRS)

**Comment by icc** [comment]
Sounds like a valid suggestion, however, I'm not very familiar with what is common activity events so I have some questions:
* Activity viewed, is that the _course_module_viewed_ event?
* I saw the activity index, but I cannot really figure out where it is used?
* I believe this would have to be some sort of xAPI event since each event type must have a separate class, and we can't create the classes on the fly. That said I'm a bit unsure who/which module would benefit from these events. Today, all event are accessible if the module adds a custom JavaScript that listen for events on the [External Event Dispatcher](https://h5p.org/events#external).
* I could be wrong but doesn't Moodle trigger the graded event automatically when `grade_update()` is used?

**Comment by nadavkav** [comment]
Great. I am already on it (As we need it in our campus, Israeli Open University) 

- Activity viewed, is that the course_module_viewed event (Yes)
- Activity index is a page where you will get a list of all the H5P activities in the course, divided into course topics. with links to those activities.
- We already added a JS event listener (see some old code here: https://github.com/h5p/h5p-moodle-plugin/issues/45 . I have some updates on that one too. I will send a PR soon) and I already implemented a some code that send the xAPI statements back into Moodle as n H5P xAPI event.
- You are right about the grades.

**Comment by icc** [comment]
Good, do you know where the link to the activity index is? I'm unable to locate it

**Comment by nadavkav** [comment]
One easy way to see it is by adding a block called "Activities" to a course. The block display a list of all Index.php pages from the activities that are included in a course.

You can also, manually, pass id=courseid to hvp/index.php (https://github.com/h5p/h5p-moodle-plugin/blob/master/index.php)

Here is a general demo for the Assignment activity:
https://qa.moodle.net/mod/assign/index.php?id=2 (user: teacher, password: test)

**Comment by icc** [comment]
Ah, thank you. I added a separate issue for fixing the index page #70.

**Comment by falcon-git** [comment]
Making any progress here @nadavkav? Hoping to release the Moodle plugin soon and would be great to have this in place!

**Comment by nadavkav** [comment]
I will upload a patch over the weekend.

Regarding the xAPI events, I am a little bit undecided (also my fellow colleagues are not sure what will be best for the long run)
I have two solution, I am not sure which one we should go forward with.
(1) Have a single hvp xAPI event - class xapi_action , which gets splitted to xapi & action in the mdl_logstore_standard_log table (which seem easy and useful for creating SQL custom reports) 
(2) Have an event for each xAPI action, like class xapi_attempted, class xapi_started ...
(Seem like more coding, but might be more elegant?)

What do you think?

**Comment by falcon-git** [comment]
I don't know much about Moodle, but there is no limit to how many different verbs you can have in xAPI statements so #2 will be a very problematic solution in the future.

**Comment by nadavkav** [comment]
I have tried both scenarios, and although using explicit event for each current known (implemented by h5p) xAPI event which looks very elegant and useful in the mdl_logstore_standard_log when considering future SQL reports... I think it would be more generic to use a single xAPI event and have all the Actor, Verb & Object data send as "other" parameters with the event (which are stored as a JSON in the "other" field of the mdl_logstore_standard_log table)

**Comment by falcon-git** [comment]
Makes sense

**Comment by nadavkav** [comment]
@garemoko , I would LOVE your input on this issue.

**Comment by garemoko** [comment]
I think there's two types of events here:

1. Standard Moodle events that are logged for all plugins. These probably don't have h5p xAPI statements and should be treated like any other Moodle events from a logstore perspective. 

2. h5P events which there is already a partial xAPI statement for which just needs some elements adding. 

In the 2nd case, since what's added to the statements should be the same each time for an H5P event, i think it makes sense to have a generic H5P event. The three different parts of the Moodle logstore plugin would then do the following:

Log expander - catch the event and fetch the base statement from an H5P plugin table referenced in the event. Collect other data needed to populate the agent and context properties needed. 

Translator - wouldn't do that much in this case; possibly some reformatting of the additional properties. 

Emitter - build and send the final statement. 

Having an event for each action seems to me unscalable and hopefully unnecessary.

**Comment by nadavkav** [comment]
Thank you Andrew (@garemoko) for you advice!
Looks perfectly logical and sensible to me.

If I can pick your brain a little bit longer...
Can you give some advice about issues #45 too

**Comment by nadavkav** [comment]
@falcon-git have you had a chance to look over the patch?

**Comment by clementpr** [comment]
Hi,

I'm not sure to understrand all of this discussions post. 
In Moodle we have differents possibilities for "activity completion". Depending on the activity (forum, test, glossary, lesson, etc.) or resources (page, link, etc.), we have that  https://docs.moodle.org/31/en/Activity_completion_settings#Activity_settings

I think with h5P, it will be better to propose one more completion activy.

**We have already**  
-  Require grade =student mus receive a grade to complete this activity (already awailable in the plugin today). This level is particulary usefull for games, questions. 
- Students can mannualy mark the activity as completed. This level is usefull to leave control to students.

**We could need too**
-  Require view : student must view this activity to complete it. 
In my view, this level is required when we just deliver simple content as "course presentation", "interactive video", etc.
With this activity completion "require view", we can offer to automatically validate an activity when student launch / see a H5P content. 

When I test the Moodle Plugin, I don't see the possibility to set "require view". Can we imagine to add this ?

![2016-09-02_14-02-19](https://cloud.githubusercontent.com/assets/8448884/18203248/ffd09404-7115-11e6-8a8d-0638b3f2653e.png)

Regards.

**Comment by thomasmars** [comment]
The initially suggested common Moodle activity events should be in place, so closing this issue. Please open a new issue for any new events that should be supported.
